src/Microsoft.ApplicationInsights/PreciseTimestamp.cs

```diff
@@ -0,0 +1,109 @@
+﻿namespace Microsoft.ApplicationInsights
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+#if NET45 || NET46
+    using System.Diagnostics;
+    using System.Threading;
+#endif
+
+    internal class PreciseTimestamp
+    {
+        private static readonly object Lck = new object();
+        private static PreciseTimestamp instance = null;
+
+#if NET45 || NET46
+        private static TimeSync timeSync;
+        private readonly Timer syncTimeUpdater;
+
+        private PreciseTimestamp()
+        {
+            this.syncTimeUpdater = InitializeSyncTimer();
+        }
+#endif
+
+        public static PreciseTimestamp Instance
```

Comment:
This seems a bit heavy-handed. Can't you just use a static auto-prop?

Reply:
No, I cannot. The problem is that I need to start a timer to correct stopwatch skew. 
Staring a timer without static constructor is a challenge - static field (unused) is optimized away. Static constructor has a perf hit, so I preferred to use singleton. 

Happy to learn if it could be avoided

Reply:
(I asked elsewhere, but I'll repeat it here.) Do you have more info on the perf hit for static ctors? I've never heard of one.

(I suppose we could measure, but it would likely take millions of iterations before a difference emerges)

Reply:
I have never done benchmarking for static constructor, there are some sources like [Jon Skeet articles](https://csharpindepth.com/articles/BeforeFieldInit) or ~~apparetly~~ [FXcop docs](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1810-initialize-reference-type-static-fields-inline)

I also agree that if there is any impact, it is likely extremely low, I'll use static constructor.